File: clients/Go/debug_interface.go
package main

import (
	"bufio"
	. "aicup2020/model"
)

type DebugInterface struct {
	Reader *bufio.Reader
	Writer *bufio.Writer
}

func (debugInterface DebugInterface) Send(command DebugCommand) {
	ClientMessageDebugMessage{
		Command: command,
	}.Write(debugInterface.Writer)
	err := debugInterface.Writer.Flush()
	if err != nil {
		panic(err)
	}
}

func (debugInterface DebugInterface) GetState() DebugState {
	ClientMessageRequestDebugState{}.Write(debugInterface.Writer)
	err := debugInterface.Writer.Flush()
	if err != nil {
		panic(err)
	}
	return ReadDebugState(debugInterface.Reader)
}

File: clients/Python-PyPy/debug_interface.py
import model


class DebugInterface:
    def __init__(self, reader, writer):
        self.reader = reader
        self.writer = writer

    def send(self, command):
        model.ClientMessage.DebugMessage(command).write_to(self.writer)
        self.writer.flush()

    def get_state(self):
        model.ClientMessage.RequestDebugState().write_to(self.writer)
        self.writer.flush()
        return model.DebugState.read_from(self.reader)
